src/daft-ir/src/proto/schema.rs

```diff
@@ -0,0 +1,371 @@
+    type Message = proto::Field;
+
+    fn from_proto(message: Self::Message) -> ProtoResult<Self>
+    where
+        Self: Sized,
+    {
+        let name = message.name;
+        let data_type = from_proto(message.data_type)?;
+        Ok(Self::new(name, data_type))
+    }
+
+    fn to_proto(&self) -> ProtoResult<Self::Message> {
+        Ok(Self::Message {
+            name: self.name.clone(),
+            data_type: self.dtype.to_proto()?.into(),
+        })
+    }
+}
+
+impl FromToProto for ir::TimeUnit {
+    type Message = i32;
+
+    fn from_proto(message: Self::Message) -> ProtoResult<Self> {
+        let time_unit = proto::TimeUnit::try_from(message).unwrap_or(proto::TimeUnit::Unspecified);
+        Ok(match time_unit {
+            proto::TimeUnit::Unspecified => Self::Nanoseconds, // default fallback
+            proto::TimeUnit::Nanoseconds => Self::Nanoseconds,
+            proto::TimeUnit::Microseconds => Self::Microseconds,
+            proto::TimeUnit::Seconds => Self::Seconds,
+        })
+    }
+
+    fn to_proto(&self) -> ProtoResult<Self::Message> {
+        let time_unit = match self {
+            Self::Nanoseconds => proto::TimeUnit::Nanoseconds,
+            Self::Microseconds => proto::TimeUnit::Microseconds,
+            Self::Milliseconds => to_proto_err!("No milliseconds!?"),
+            Self::Seconds => proto::TimeUnit::Seconds,
+        };
+        Ok(time_unit as i32)
+    }
+}
+
+impl FromToProto for ir::ImageMode {
+    type Message = i32;
+
+    fn from_proto(message: Self::Message) -> ProtoResult<Self> {
+        let image_mode =
+            proto::ImageMode::try_from(message).unwrap_or(proto::ImageMode::Unspecified);
+        Ok(match image_mode {
+            proto::ImageMode::Unspecified => Self::L, // default fallback
+            proto::ImageMode::L => Self::L,
+            proto::ImageMode::La => Self::LA,
+            proto::ImageMode::Rgb => Self::RGB,
+            proto::ImageMode::Rgba => Self::RGBA,
+            proto::ImageMode::L16 => Self::L16,
+            proto::ImageMode::La16 => Self::LA16,
+            proto::ImageMode::Rgb16 => Self::RGB16,
+            proto::ImageMode::Rgba16 => Self::RGBA16,
+            proto::ImageMode::Rgb32f => Self::RGB32F,
+            proto::ImageMode::Rgba32f => Self::RGBA32F,
+        })
+    }
+
+    fn to_proto(&self) -> ProtoResult<Self::Message> {
+        let image_mode = match self {
+            Self::L => proto::ImageMode::L,
+            Self::LA => proto::ImageMode::La,
+            Self::RGB => proto::ImageMode::Rgb,
+            Self::RGBA => proto::ImageMode::Rgba,
+            Self::L16 => proto::ImageMode::L16,
+            Self::LA16 => proto::ImageMode::La16,
+            Self::RGB16 => proto::ImageMode::Rgb16,
+            Self::RGBA16 => proto::ImageMode::Rgba16,
+            Self::RGB32F => proto::ImageMode::Rgb32f,
+            Self::RGBA32F => proto::ImageMode::Rgba32f,
+        };
+        Ok(image_mode as i32)
+    }
+}
```

Comment:
we should also add some rust tests in each of these modules to ensure that the round trip serialization works.